pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -0,0 +1,389 @@
+	if err != nil {
+		return err
+	}
+
+	r.(*ReconcileClusterPool).dynamicClient, err = dynamic.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return err
+	}
+
+	r.(*ReconcileClusterPool).discoveryClient, err = discovery.NewDiscoveryClientForConfig(mgr.GetConfig())
+	if err != nil {
+		return err
+	}
+
+	// TODO: watch ClusterDeployment, map to it's owning Pool (if any) and queue it up
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileClusterPool{}
+
+// ReconcileClusterPool reconciles a CLusterLeasePool object
+type ReconcileClusterPool struct {
+	client.Client
+	scheme          *runtime.Scheme
+	dynamicClient   dynamic.Interface
+	discoveryClient discovery.DiscoveryInterface
+}
+
+// Reconcile reads that state ClusterPool, checks if we currently have enough ClusterDeployments waiting, and
+// attempts to reach the desired state if not.
+func (r *ReconcileClusterPool) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	start := time.Now()
+	logger := log.WithFields(log.Fields{
+		"clusterPool": request.Name,
+		"controller":  ControllerName,
+	})
+
+	logger.Infof("reconciling cluster pool: %v", request.Name)
+	defer func() {
+		dur := time.Since(start)
+		hivemetrics.MetricControllerReconcileTime.WithLabelValues(ControllerName).Observe(dur.Seconds())
+		logger.WithField("elapsed", dur).Info("reconcile complete")
+	}()
+
+	// Fetch the ClusterPool instance
+	clp := &hivev1.ClusterPool{}
+	err := r.Get(context.TODO(), request.NamespacedName, clp)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("pool not found")
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		log.WithError(err).Error("error reading cluster pool")
+		return reconcile.Result{}, err
+	}
+
+	// If the pool is deleted, do not reconcile.
+	if clp.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	// Find all ClusterDeployments from this pool:
+	allCDs := &hivev1.ClusterDeploymentList{}
+	if err := r.Client.List(context.Background(), allCDs); err != nil {
+		logger.WithError(err).Error("error listing ClusterDeployments")
+		return reconcile.Result{}, err
+	}
+	var poolCDs []hivev1.ClusterDeployment
+	for _, cd := range allCDs.Items {
+		if cd.Spec.ClusterPoolRef != nil && cd.Spec.ClusterPoolRef.Name == clp.Name && cd.Spec.ClusterPoolRef.Namespace == clp.Namespace {
+			poolCDs = append(poolCDs, cd)
+		}
+	}
+
+	installing := 0
+	deleting := 0
+	for _, cd := range poolCDs {
+		if cd.DeletionTimestamp != nil {
+			deleting += 1
+		} else if !cd.Spec.Installed {
+			installing += 1
+		}
+	}
+	logger.WithFields(log.Fields{
+		"installing": installing,
+		"deleting":   deleting,
+		"total":      len(poolCDs),
+		"ready":      len(poolCDs) - installing - deleting,
+	}).Info("found clusters for ClusterPool")
+
+	// If too many, delete some.
```

Comment:
We also need to handle replacing unclaimed ClusterDeployments when the spec of the ClusterPool changes. For example, if the base domain is changed, then all of the existing, unclaimed ClusterDeployments need to be deleted and replaced with new ones.

Reply:
This is something I was hoping to avoid, changing the release image was the most common case I could think of but I was hoping we could ignore it and let it only affect new clusters being spawned. This would match what we see with the machine API for example, but we hate that, so you're probably right.

I assume we're talking about hashing the spec, storing in status, and if we see a change wipe out all clusters.

I will add a Jira, I'd like to defer this though and get this PR in.

Reply:
I think if a user is changing the pool, they don't want to continue getting clusters with the old specs until the pre-existing clusters are all claimed. In practice, I would expect a user that discovers the behavior to always reduce the pool size to 0 and increase it back to what they want to cycle the pool. It seems onerous to put that burden on the user rather than having Hive do it itself.

I am fine with deferring.